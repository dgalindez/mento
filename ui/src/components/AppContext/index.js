import { createContext, useEffect } from 'react';

import useGeolocation from '../../hooks/useGeolocation';
import useLocale from '../../hooks/useLocale';
import useWeatherApi from '../../hooks/useWeatherApi';

const AppContext = createContext({
  geolocation: {},
  locale: {},
  weather: {},
});

export const ContextComponent = ({ children }) => {
  const geolocation = useGeolocation();
  const locale = useLocale();
  const weather = useWeatherApi();

  useEffect(() => {
    if (!geolocation?.failed) {
      weather?.getLocationData(geolocation.lat, geolocation.lon);
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [geolocation.failed, geolocation.lat, geolocation.lon]);

  return (
    <AppContext.Provider value={{ geolocation, locale, weather }}>
      {children}
    </AppContext.Provider>
  );
};

export default AppContext;
