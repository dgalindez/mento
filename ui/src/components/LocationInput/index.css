.location {
  width: 200px;
}
