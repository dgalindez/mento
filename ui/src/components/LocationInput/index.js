import { useContext, useState } from 'react';
import { useTranslation } from 'react-i18next';

import AppContext from '../AppContext';
import { UNITS } from '../../utils/constants';

const LocationInput = () => {
  const { t } = useTranslation();
  const {
    locale: {
      locale,
      setLocale,
      setUnits,
      setZip,
      units,
      zip,
    },
    weather: {
      getZipData,
    },
  } = useContext(AppContext);

  // Used to avoid re-rendering without re-fetching
  const [localUnits, setLocalUnits] = useState(units);

  const changeLocale = ({ target: { value } = {} } = {}) => setLocale(value);
  const changeZip = ({ target: { value } = {} } = {}) => setZip(value);
  const changeUnits = ({ target: { value } = {} } = {}) => setLocalUnits(value);

  const doSearch = () => {
    setUnits(localUnits);
    getZipData(zip, locale, localUnits);
  };

  return (
    <>
      <div className="columns m-4 mx-6">
        <div className="field column is-one-quarter">
          <label className="label">{t('input.zipLabel')}</label>
          <div className="control">
            <input
              className="input"
              placeholder={t('input.zipLabel')}
              type="text"
              value={zip}
              onChange={changeZip}
            />
          </div>
        </div>
        <div className="field column is-one-quarter">
          <label className="label">{t('input.localeLabel')}</label>
          <div className="control">
            <input
              className="input"
              placeholder={t('input.localeLabel')}
              type="text"
              value={locale}
              onChange={changeLocale}
            />
          </div>
        </div>
        <div className="field column is-one-quarter">
          <label className="label">{t('input.unitsLabel')}</label>
          <div className="control">
            <div className="select">
              <select onChange={changeUnits} value={localUnits}>
                {Object.keys(UNITS).map(unit => (
                  <option key={UNITS[unit]}>
                    {t(`input.select.${UNITS[unit]}`)}
                  </option>
                ))}
              </select>
            </div>
          </div>
        </div>
      </div>
      <button
        className="button is-medium is-success ml-6 mb-6"
        onClick={doSearch}
        disabled={!zip || !locale}
      >
        {t('input.button')}
      </button>
    </>
  );
};

export default LocationInput;
