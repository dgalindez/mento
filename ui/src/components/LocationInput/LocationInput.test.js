import { render, screen, waitFor } from '@testing-library/react';import userEvent from '@testing-library/user-event';
import { I18nextProvider } from 'react-i18next';

import i18n from '../../utils/i18nTestHelper';
import AppContext from '../AppContext';

import LocationInput from '.';

describe('LocationInput', () => {
  let contextData = {};

  beforeEach(() => {
    contextData = {
      geolocation: {
        loading: false,
        failed: false,
      },
      locale: {
        setLocale: jest.fn(),
        setUnits: jest.fn(),
        setZip: jest.fn(),
      },
      weather: {
        error: '',
        loading: false,
        getZipData: jest.fn(),
      },
    }
  });

  it('renders correctly', () => {
    const { container: view } = render(
      <I18nextProvider i18n={i18n}>
        <AppContext.Provider value={contextData}>
          <LocationInput />
        </AppContext.Provider>
      </I18nextProvider>
    );

    expect(view).toMatchSnapshot();
  });

  it('updates zip and locale', async () => {
    render(
      <I18nextProvider i18n={i18n}>
        <AppContext.Provider value={contextData}>
          <LocationInput />
        </AppContext.Provider>
      </I18nextProvider>
    );

    userEvent.type(
      screen.getByPlaceholderText('input.zipLabel'),
      'zip',
      { delay: 1 }
    );

    await waitFor(
      () => expect(contextData.locale.setZip).toHaveBeenCalledWith('zip')
    );
    expect(screen.getByPlaceholderText('input.zipLabel')).toMatchSnapshot();

    userEvent.type(
      screen.getByPlaceholderText('input.localeLabel'),
      'locale',
      { delay: 1 }
    );

    await waitFor(
      () => expect(contextData.locale.setLocale).toHaveBeenCalledWith('locale')
    );
    expect(screen.getByPlaceholderText('input.localeLabel')).toMatchSnapshot();
  });

  it('updates selected units', async () => {
    render(
      <I18nextProvider i18n={i18n}>
        <AppContext.Provider value={contextData}>
          <LocationInput />
        </AppContext.Provider>
      </I18nextProvider>
    );

    userEvent.click(screen.getByText('input.select.imperial'));
    userEvent.click(screen.getByText('input.select.metric'));

    expect(screen.getByText('input.select.metric')).toMatchSnapshot();
  });

  it('performs search', async () => {
    render(
      <I18nextProvider i18n={i18n}>
        <AppContext.Provider value={contextData}>
          <LocationInput />
        </AppContext.Provider>
      </I18nextProvider>
    );


    userEvent.type(
      screen.getByPlaceholderText('input.zipLabel'),
      'zip',
      { delay: 1 }
    );

    userEvent.type(
      screen.getByPlaceholderText('input.localeLabel'),
      'locale',
      { delay: 1 }
    );

    userEvent.click(screen.getByText('input.button'));
    expect(contextData.weather.getZipData).toHaveBeenCalledTimes(0);
  });
});
