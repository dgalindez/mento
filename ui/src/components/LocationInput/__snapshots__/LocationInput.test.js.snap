// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LocationInput renders correctly 1`] = `
<div>
  <div
    class="columns m-4 mx-6"
  >
    <div
      class="field column is-one-quarter"
    >
      <label
        class="label"
      >
        input.zipLabel
      </label>
      <div
        class="control"
      >
        <input
          class="input"
          placeholder="input.zipLabel"
          type="text"
          value=""
        />
      </div>
    </div>
    <div
      class="field column is-one-quarter"
    >
      <label
        class="label"
      >
        input.localeLabel
      </label>
      <div
        class="control"
      >
        <input
          class="input"
          placeholder="input.localeLabel"
          type="text"
          value=""
        />
      </div>
    </div>
    <div
      class="field column is-one-quarter"
    >
      <label
        class="label"
      >
        input.unitsLabel
      </label>
      <div
        class="control"
      >
        <div
          class="select"
        >
          <select>
            <option>
              input.select.standard
            </option>
            <option>
              input.select.metric
            </option>
            <option>
              input.select.imperial
            </option>
          </select>
        </div>
      </div>
    </div>
  </div>
  <button
    class="button is-medium is-success ml-6 mb-6"
    disabled=""
  >
    input.button
  </button>
</div>
`;

exports[`LocationInput updates selected units 1`] = `
<option>
  input.select.metric
</option>
`;

exports[`LocationInput updates zip and locale 1`] = `
<input
  class="input"
  placeholder="input.zipLabel"
  type="text"
  value=""
/>
`;

exports[`LocationInput updates zip and locale 2`] = `
<input
  class="input"
  placeholder="input.localeLabel"
  type="text"
  value=""
/>
`;
